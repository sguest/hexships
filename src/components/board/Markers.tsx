import { useEffect, useRef } from 'react';
import UiSettings from '../../config/UiSettings';
import { Point } from '../../utils/point-utils';
import * as hexUtils from '../../utils/hex-utils';
import * as pointUtils from '../../utils/point-utils';

export interface MarkersProps {
    hits?: Point[]
    misses?: Point[]
    uiSettings: UiSettings
}

export default function Markers(props: MarkersProps) {
    const canvasRef = useRef<HTMLCanvasElement>(null);

    useEffect(() => {
        const canvas = canvasRef.current;
        if(!canvas) {
            return;
        }
        const context = canvas.getContext('2d');
        if(!context) {
            return;
        }

        context.clearRect(0, 0, canvas.width, canvas.height);

        const drawMarkers = (markers: Point[] | undefined, colour: string) => {
            if(markers) {
                context.strokeStyle = 'black';
                context.fillStyle = colour;
                for(const marker of markers) {
                    context.beginPath();
                    const coords = pointUtils.add(hexUtils.getCenter(marker, props.uiSettings.cellSize), props.uiSettings.gridOffset);
                    context.arc(coords.x, coords.y, props.uiSettings.cellSize * 0.4, 0, Math.PI * 2);
                    context.stroke();
                    context.fill();
                }
            }
        }

        drawMarkers(props.hits, 'red');
        drawMarkers(props.misses, 'white');
    }, [props.hits, props.misses, props.uiSettings, canvasRef])

    return <canvas ref={canvasRef} width="500" height="500" className="markers-canvas" />
}