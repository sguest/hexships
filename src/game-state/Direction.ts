enum Direction {
    positiveX,
    positiveY,
    positiveZ,
    negativeX,
    negativeY,
    negativeZ,
}

export default Direction

const deltas = {
    [Direction.positiveX]: { x: 1, y: 0 },
    [Direction.positiveY]: { x: 0, y: 1 },
    [Direction.positiveZ]: { x: -1, y: 1 },
    [Direction.negativeX]: { x: -1, y: 0 },
    [Direction.negativeY]: { x: 0, y: -1 },
    [Direction.negativeZ]: { x: 1, y: -1 },
}

export function getDelta(direction: Direction): { x: number, y: number } {
    return deltas[direction];
}

export function allDirections() {
    return [Direction.positiveX, Direction.positiveY, Direction.positiveZ, Direction.negativeX, Direction.negativeY, Direction.negativeZ];
}