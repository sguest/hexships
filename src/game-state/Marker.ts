import { Point } from '../utils/point-utils';

export enum MarkerType {
    Hit,
    Miss,
}

export default interface Marker extends Point {
    type: MarkerType
}