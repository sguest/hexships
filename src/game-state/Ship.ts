import Direction from './Direction';
import * as direction from './Direction';
import * as pointUtils from '../utils/point-utils';
import * as hexUtils from '../utils/hex-utils';

export default interface Ship {
    x: number
    y: number
    size: number
    facing: Direction
    hits: number
    name: string
}

export function getPoints(ship: Ship) {
    const delta = direction.getDelta(ship.facing);
    let point = { x: ship.x, y: ship.y };
    const points = [{ x: ship.x, y: ship.y }];
    for(let i = 1; i < ship.size; i++) {
        point = pointUtils.add(point, delta);
        points.push(point);
    }
    return points;
}

export function isPlacementValid(ships: Ship[], gridSize: number) {
    const points = ships.map(s => getPoints(s));
    for(let i = 0; i < ships.length; i++) {
        for(const point of points[i]) {
            if(!hexUtils.isInGrid(point, gridSize)) {
                return false;
            }
            for(let j = i + 1; j < ships.length; j++) {
                for(const point2 of points[j]) {
                    if(pointUtils.equal(point, point2)) {
                        return false;
                    }
                }
            }
        }
    }

    return true;
}