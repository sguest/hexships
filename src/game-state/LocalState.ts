import { Point } from "../utils/point-utils";
import Ship from "./Ship";

export default interface LocalState {
    ownShips: Ship[];
    ownHits: Point[];
    ownMisses: Point[];
    opponentHits: Point[];
    opponentMisses: Point[];
}