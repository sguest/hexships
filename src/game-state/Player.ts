import { Point } from "../utils/point-utils";
import Ship from "./Ship";

export default interface Player {
    ships: Ship[]
    hits: Point[]
    misses: Point[]
}