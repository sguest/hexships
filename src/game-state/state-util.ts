import GameSettings from '../config/GameSettings';
import LocalState from './LocalState';
import Ship from './Ship';

export function getSunkShips(ships: Ship[]) {
    return ships.filter(s => s.hits === s.size);
}

export function getNumShots(settings: GameSettings, state: LocalState) {
    if(settings.shotPerShip) {
        return state.ownShips.length - getSunkShips(state.ownShips).length;
    }
    else {
        return settings.shots;
    }
}