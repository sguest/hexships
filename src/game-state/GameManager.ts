import GameSettings from '../config/GameSettings';
import LocalState from './LocalState';
import Player from './Player';
import Ship, * as shipFuncs from './Ship';
import { Point } from '../utils/point-utils';
import * as pointUtils from '../utils/point-utils';
import * as hexUtils from '../utils/hex-utils';
import { MarkerType } from './Marker';

export default class GameManager {
    private players: Player[];
    private activePlayerId: number;

    constructor(public gameSettings: GameSettings, private subscriber: (playerId: number, state: LocalState) => void) {
        this.players = [
            {
                ships: [],
                markers: [],
            },
            {
                ships: [],
                markers: [],
            },
        ];
        this.activePlayerId = 0;
    }

    public getLocalState(playerId: number): LocalState {
        const otherPlayerId = +!playerId;
        return {
            ownShips: this.players[playerId].ships,
            ownMarkers: this.players[playerId].markers,
            opponentMarkers: this.players[otherPlayerId].markers,
            isOwnTurn: playerId === this.activePlayerId && !!this.players[playerId].ships.length,
            sunkEnemies: this.players[otherPlayerId].ships.filter(s => s.hits === s.size).map(s => s.name),
        }
    }

    private validateShipPlacement(ships: Ship[]) {
        if(!shipFuncs.isPlacementValid(ships, this.gameSettings.gridSize)) {
            return [];
        }

        const neededShips = this.gameSettings.ships.slice(0);
        const newShips: Ship[] = [];

        for(const ship of ships) {
            const index = neededShips.findIndex(s => s.name === ship.name && s.size === ship.size);
            if(index === -1) {
                return [];
            }

            newShips.push({
                x: ship.x,
                y: ship.y,
                size: ship.size,
                facing: ship.facing,
                hits: 0,
                name: ship.name,
            });

            neededShips.splice(index, 1);
        }

        if(neededShips.length) {
            return [];
        }

        return newShips;
    }

    public setShips(playerId: number, ships: Ship[]) {
        const newShips = this.validateShipPlacement(ships);
        if(newShips.length) {
            this.players[playerId].ships = newShips;
        }
        this.broadcastState();
    }

    public fireShot(playerId: number, target: Point) {
        const valid = this.activePlayerId === playerId &&
            hexUtils.isInGrid(target, this.gameSettings.gridSize) &&
            !this.players[playerId].markers.some(marker => pointUtils.equal(marker, target));

        if(valid) {
            const otherPlayerId = +!playerId;
            let isHit = false;
            for(const ship of this.players[otherPlayerId].ships) {
                for(const point of shipFuncs.getPoints(ship)) {
                    if(pointUtils.equal(point, target)) {
                        isHit = true;
                        ship.hits++;
                    }
                }
            }
            this.players[playerId].markers.push({
                x: target.x,
                y: target.y,
                type: isHit ? MarkerType.Hit : MarkerType.Miss,
            })
            this.activePlayerId = otherPlayerId;
        }
        this.broadcastState();
    }

    private broadcastState() {
        for(let id = 0; id < this.players.length; id++) {
            this.subscriber(id, this.getLocalState(id));
        }
    }
}