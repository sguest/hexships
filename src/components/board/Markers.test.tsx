import 'jest-canvas-mock';
import { render, screen } from '@testing-library/react'
import { MarkerType } from '../../game-state/Marker'
import Markers from './Markers'

test('Should render markers', () => {
    const markers = [
        { x: 3, y: 2, type: MarkerType.Hit },
        { x: -2, y: 1, type: MarkerType.Miss },
    ]
    render(<Markers markers={markers} uiScale={1} gridDimensions={{ x: 1, y: 1 }} gridSize={7} />)
    const canvas = screen.getByTestId('marker-canvas') as HTMLCanvasElement;
    const context = canvas.getContext('2d');
    const events = context!.__getEvents();
    expect(events).toMatchSnapshot();
})