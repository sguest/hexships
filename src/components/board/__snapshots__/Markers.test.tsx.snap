// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render markers 1`] = `
Array [
  Object {
    "props": Object {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": "#ff0000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 8,
      "startAngle": 0,
      "x": 90.5,
      "y": 121.7435565298214,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 8,
            "startAngle": 0,
            "x": 90.5,
            "y": 121.7435565298214,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 8,
            "startAngle": 0,
            "x": 90.5,
            "y": 121.7435565298214,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 8,
      "startAngle": 0,
      "x": -59.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 8,
            "startAngle": 0,
            "x": -59.5,
            "y": 0.5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 8,
            "startAngle": 0,
            "x": -59.5,
            "y": 0.5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;
